runtime/interpreter/interpreter_expression.go

```diff
@@ -1142,7 +1184,66 @@ func (interpreter *Interpreter) VisitPathExpression(expression *ast.PathExpressi
 	)
 }
 
-func (interpreter *Interpreter) VisitAttachExpression(_ *ast.AttachExpression) Value {
-	// TODO: implement this
-	return nil
+func (interpreter *Interpreter) VisitAttachExpression(attachExpression *ast.AttachExpression) Value {
+
+	base, ok := interpreter.evalExpression(attachExpression.Base).(*CompositeValue)
+	// we enforce this in the checker
+	if !ok {
+		panic(errors.NewUnreachableError())
+	}
+
+	locationRange := LocationRange{
+		Location:    interpreter.Location,
+		HasPosition: attachExpression,
+	}
+
+	if inIteration := interpreter.SharedState.inAttachmentIteration[base]; inIteration {
+		panic(AttachmentIterationMutationError{
+			Value:         base,
+			LocationRange: locationRange,
+		})
+	}
+
+	// the `base` value must be accessible during the attachment's constructor, but we cannot
+	// set it on the attachment's `CompositeValue` yet, because the value does not exist. Instead
+	// we save the base value in the interpreter's shared state and set the variable directly inside the
+	// constructor to make it available
+	baseValue := NewEphemeralReferenceValue(
+		interpreter,
+		false,
+		base,
+		interpreter.MustSemaTypeOfValue(base).(*sema.CompositeType),
+	)
+	interpreter.trackReferencedResourceKindedValue(base.StorageID(), base)
+
+	interpreter.SharedState.deferredBaseValue = baseValue
+	attachment, ok := interpreter.VisitInvocationExpression(attachExpression.Attachment).(*CompositeValue)
+
+	// Because `self` in attachments is a reference, we need to track the attachment if it's a resource
+	interpreter.trackReferencedResourceKindedValue(attachment.StorageID(), attachment)
+
+	base = base.Transfer(
+		interpreter,
+		locationRange,
+		atree.Address{},
+		false,
+		nil,
+	).(*CompositeValue)
+
+	// we enforce this in the checker
+	if !ok {
+		panic(errors.NewUnreachableError())
+	}
+
+	// when `v[A]` is executed, we set `A`'s base to `&v`
+	attachment.setBaseValue(interpreter, base)
```

Comment:
Can't we directly pass the `baseValue` here,  instead of passing `base` and creating another `EphemeralReferenceValue` inside it?

Reply:
This function is used elsewhere and only takes a `CompositeValue`. To change this to take a reference value would require all the call sites to construct their own reference values instead of having the logic encapsulated inside this function.